README.md
# RepoSmallBadge

[![Gem Version](https://badge.fury.io/rb/repo-small-badge.svg)](https://badge.fury.io/rb/repo-small-badge)
[![Build Status](https://api.travis-ci.org/MarcGrimme/repo-small-badge.svg?branch=master)](https://secure.travis-ci.org/MarcGrimme/repo-small-badge)
[![Depfu](https://badges.depfu.com/badges/e794da00404482f66ce7ca05eea9640b/count.svg)](https://depfu.com/github/MarcGrimme/repo-small-badge?project_id=6915)
![Coverage](https://marcgrimme.github.io/repo-small-badge/badges/coverage_badge_total.svg)

*RepoSmalladge* is a gem that can be added to the `Gemfile` and consolidates functionality to produce a badge file.
The text place and other configurable can be specified at instance time.

## Installation

The badge creation is dependent on the [Victor gem](https://github.com/DannyBen/victor) for creating the SVG formatted badge.

It can be installed in your Ruby library or rails app as part of the `Gemfile` as follows.

```
# In your gemfile
gem 'repo-small-badge', :require => false
```

## Usage

Most propably you need a surrounding formatter for a different badge you want to create and this gem is implicitly required.

Examples are:
* [SimpleCovSmallBadge](https://github.com/MarcGrimme/simplecov-small-badge)

## Configuration Options

The behaviour of `RepoSmallBadge` can be influenced by configuration options that are passed to the constructor as a hash.
The following options are supported:

## Development

After checking out the repo, run `bundle update` to install dependencies. Then, run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release` to create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

1. Fork it ( https://github.com/marcgrimme/repo-small-badge/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
# Change Logs

# 0.2.4

* Fix format SVG to look better
* Support rubycritic badge (perliminary)

# 0.2.3

* Support rubycritic
* More flexibility for configuring the different font colors and sizes.

# 0.2.2

* Fix documentation to refer to victor instead of Minimagic
* Fix travis build to produce a bagde


spec/spec_helper.rb
# frozen_string_literal: true

require 'simplecov'
require 'byebug'

SimpleCov.start do
  module SimpleCovSmallBadge
    class Formatter
    end
  end

  SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter.new(
    [
      SimpleCov::Formatter::HTMLFormatter,
      SimpleCovSmallBadge::Formatter
    ]
  )
end

SimpleCov.minimum_coverage 100

# require 'simplecov-small-badge'
require 'repo_small_badge'
require 'simplecov_small_badge'
require 'rubygems'
require 'bundler/setup'

Dir[File.join('./spec/support/*.rb')].each { |f| require f }

SimpleCovSmallBadge.configure do |config|
  config.format = 'svg'
end
RSpec.configure do |config|
  # some (optional) config here
end


spec/support/mocks.rb
# frozen_string_literal: true

module TestRepoSmallBadge
  # rubocop:disable Metrics/MethodLength,Metrics/LineLength,Metrics/ParameterLists
  def rounded_svg_string(title_color: '#fff', title_font: 'Verdana,sans-serif', title_font_size: 11,
                         value_color: '#fff', value_font: 'Verdana,sans-serif', value_font_size: 11)
    %(<svg contentScriptType="text/ecmascript" contentStyleType="text/css" preserveAspectRatio="xMidYMid meet" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" version="1.0" height="20" width="120">

<linearGradient id="smooth" x2="0" y2="120">
<stop offset="0" stop-color="#bbb" stop-opacity=".1"/>
<stop offset="1" stop-opacity=".1"/>
</linearGradient>
<clipPath id="round">
<rect height="20" width="120" rx="3" fill="#fff"/>
</clipPath>
<g clip-path="url(#round)">
<rect height="20" width="60" fill="#555"/>
<rect x="60" height="20" width="60" fill="#4c1"/>
<rect height="20" width="120" fill="url(#smooth)"/>
</g>
<g fill="#{title_color}" text-anchor="middle" font-family="#{title_font}" font-size="#{title_font_size}">
<text x="30" y="15" fill="#010101" fill-opacity="0.3">
Total
</text>
<text x="30" y="14">
Total
</text>
</g>
<g fill="#{value_color}" text-anchor="middle" font-family="#{value_font}" font-size="#{value_font_size}">
<text x="90" y="15" fill="#010101" fill-opacity="0.3">
100%
</text>
<text x="90" y="14">
100%
</text>
</g>
</svg>)
  end

  def not_rounded_svg_string
    %(<svg contentScriptType="text/ecmascript" contentStyleType="text/css" preserveAspectRatio="xMidYMid meet" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" version="1.0" height="20" width="200">

<linearGradient id="smooth" x2="0" y2="200">
<stop offset="0" stop-color="#bbb" stop-opacity=".1"/>
<stop offset="1" stop-opacity=".1"/>
</linearGradient>
<clipPath id="round">
<rect height="20" width="200" rx="0" fill="#fff"/>
</clipPath>
<g clip-path="url(#round)">
<rect height="20" width="100" fill="#555"/>
<rect x="100" height="20" width="100" fill="#4c1"/>
<rect height="20" width="200" fill="url(#smooth)"/>
</g>
<g fill="#fff" text-anchor="middle" font-family="Verdana,sans-serif" font-size="11">
<text x="50" y="15" fill="#010101" fill-opacity="0.3">
badge Total
</text>
<text x="50" y="14">
badge Total
</text>
</g>
<g fill="#fff" text-anchor="middle" font-family="Verdana,sans-serif" font-size="11">
<text x="150" y="15" fill="#010101" fill-opacity="0.3">
100%
</text>
<text x="150" y="14">
100%
</text>
</g>
</svg>)
  end
  # rubocop:enable Metrics/MethodLength,Metrics/LineLength,Metrics/ParameterLists
end


spec/repo_small_badge_spec.rb
# frozen_string_literal: true

require 'spec_helper'

describe RepoSmallBadge::Image do
  include TestRepoSmallBadge

  context '#coverage_total' do
    describe '#coverage' do
      it do
        allow(File).to receive(:write)
          .with('./badge_total.svg', default_svg_string).and_return(true)
        expect(subject.badge('total', 'Total', '100%'))
          .to be_truthy
      end

      context 'without rounded and width 200 and title' do
        subject do
          described_class
            .new(badge_width: 200, rounded_border: false, title_prefix: 'badge')
        end

        it do
          allow(File).to receive(:write)
            .with('./badge_total.svg', not_rounded_svg_string).and_return(true)
          expect(subject.badge('total', 'Total', '100%')).to be_truthy
        end
      end

      context '#config_merge' do
        it do
          expect(subject.config_merge(badge_height: 10)[:badge_height]).to eq 10
        end
      end
    end
  end
end


lib/repo_small_badge/image.rb
# frozen_string_literal: true

require 'victor'
require 'repo_small_badge/configuration'

module RepoSmallBadge
  # Class that creates the Badge SVG image.
  # Behaviour is defined by the configuration hash.
  # For the different configurables see RepoSmallBadge::Configuration.
  class Image < Victor::SVGBase
    # Create new instance.
    # @config is a Hash of configurables. Keys are symbols.
    def initialize(config = {})
      @config = Configuration.new(config)
      super(template: :html, contentScriptType: 'text/ecmascript',
            contentStyleType: 'text/css', preserveAspectRatio: 'xMidYMid meet',
            'xmlns:xlink': 'http://www.w3.org/1999/xlink',
            xmlns: 'http://www.w3.org/2000/svg', version: '1.0',
            height: @config.badge_height, width: @config.badge_width)
    end

    # Creates the badge.
    # @name the suffix for the filename (badge_#{name})
    # @title the title of the badge.
    # @value is the overall value to be written to the right side of the badge.
    def badge(name, title, value)
      svg_header
      svg_rounded_box
      svg_boxes
      svg_title(title)
      svg_value(value)
      save(@config.filename(name))
    end

    # Updates the configuration settings and overwrites existing ones.
    # @config the hash that has to be merged.
    def config_merge(config)
      @config.merge!(config)
    end

    private

    def svg_header
      element :linearGradient, id: 'smooth', x2: '0', y2: @config.badge_width do
        element :stop, offset: '0', 'stop-color': '#bbb', 'stop-opacity': '.1'
        element :stop, offset: '1', 'stop-opacity': '.1'
      end
    end

    # rubocop:disable Metrics/MethodLength
    def svg_boxes
      element :g, 'clip-path' => 'url(#round)' do
        element :rect, height: @config.badge_height,
                       width: @config.badge_middle,
                       fill: @config.title_background
        element :rect, x: @config.badge_middle,
                       height: @config.badge_height,
                       width: @config.badge_middle,
                       fill: @config.value_background
        element :rect, height: @config.badge_height,
                       width: @config.badge_width, fill: 'url(#smooth)'
      end
      # rubocop:enable Metrics/MethodLength
    end

    def svg_rounded_box
      element :clipPath, id: 'round' do
        element :rect, height: @config.badge_height, width: @config.badge_width,
                       rx: @config.rounded_edge_radius, fill: @config.background
      end
    end

    # rubocop:disable Metrics/AbcSize
    def svg_title(title)
      element :g, fill: @config.title_color, 'text-anchor': 'middle',
                  'font-family': @config.title_font,
                  'font-size': @config.title_font_size do |_svg|
        element :text, @config.title(title),
                x: @config.badge_middle / 2, y: @config.badge_height - 5,
                fill: '#010101', 'fill-opacity': '0.3'
        element :text, @config.title(title),
                x: @config.badge_middle / 2, y: @config.badge_height - 6
      end
    end

    # rubocop:disable Metrics/MethodLength
    def svg_value(value)
      element :g, fill: @config.value_color, 'text-anchor': 'middle',
                  'font-family': @config.value_font,
                  'font-size': @config.value_font_size do |_svg|
        element :text, value,
                x: @config.badge_middle / 2 + @config.badge_middle,
                y: @config.badge_height - 5,
                fill: '#010101', 'fill-opacity': '0.3'
        element :text, value,
                x: @config.badge_middle / 2 + @config.badge_middle,
                y: @config.badge_height - 6
      end
    end
    # rubocop:enable Metrics/MethodLength
    # rubocop:enable Metrics/AbcSize
  end
end
